pkg/client/kubernetes/client.go

```diff
@@ -261,9 +262,16 @@ func newClientSet(conf *Config) (Interface, error) {
 
 	var runtimeClient client.Client
 	if UseCachedRuntimeClients && !conf.disableCachedClient {
-		runtimeClient, err = newRuntimeClientWithCache(conf.restConfig, conf.clientOptions, runtimeCache)
-		if err != nil {
-			return nil, err
+		if len(conf.specificallyCachedObjects) > 0 {
+			runtimeClient, err = utils.NewClientWithSpecificallyCachedReader(runtimeCache, conf.restConfig, conf.clientOptions, conf.readSpecifiedFromCache, conf.specificallyCachedObjects...)
```

Comment:
Btw. what is the reason we create an entirely new client object for the cached client instead of just creating the cache and use it together with the `directClient` in `client.DelegatingClient`? It hasn't changed with this PR but still wanted to ask for the motivation.

Reply:
TBH: I just implemented it the same way, c-r does it ([ref](https://github.com/kubernetes-sigs/controller-runtime/blob/ac380d61764a160b32946e606b0c9ecd2834e3e8/pkg/manager/manager.go#L374)) and haven't given this much thought.

Maybe it would make sense to reuse the direct client, but there might have been a concrete reason, why c-r implemented it like this.
Also I wouldn't be comfortable doing this change right now actually, because it will foster lock contention in the single client even more, which might be really bad for large landscapes (see similar issues with the RESTMapper). I still want to open an issue in c-r about the lock contention and do some benchmarks/profiling to have some numbers/proof. This will hopefully give me more confidence about making a decision here and I can reiterate then :)

Reply:
Okay, got it. Thanks for the explanation.